Verizon-GO_GENERIC_LIB/doc/retry-configuration.md

# RetryConfiguration

The following parameters are configurable for the RetryConfiguration:

## Properties

| Name | Type | Description | Setter | Getter |
|  --- | --- | --- | --- | --- |
| maxRetryAttempts | `int64` | Maximum number of retries.<br>*Default*: `0` | `WithMaxRetryAttempts` | `MaxRetryAttempts()` |
| retryOnTimeout | `bool` | Whether to retry on request timeout.<br>*Default*: `true` | `WithRetryOnTimeout` | `RetryOnTimeout()` |
| retryInterval | `time.Duration` | Interval before next retry. Used in calculation of wait time for next request in case of failure.<br>*Default*: `1` | `WithRetryInterval` | `RetryInterval()` |
| maximumRetryWaitTime | `time.Duration` | Overall wait time for the requests getting retried.<br>*Default*: `0` | `WithMaximumRetryWaitTime` | `MaximumRetryWaitTime()` |
| backoffFactor | `int64` | Used in calculation of wait time for next request in case of failure.<br>*Default*: `2` | `WithBackoffFactor` | `BackoffFactor()` |
| httpStatusCodesToRetry | `[]int64` | Http status codes to retry against.<br>*Default*: `[]int64{408, 413, 429, 500, 502, 503, 504, 521, 522, 524}` | `WithHttpStatusCodesToRetry` | `HttpStatusCodesToRetry()` |
| httpMethodsToRetry | `[]string` | Http methods to retry against.<br>*Default*: `[]string{"GET", "PUT"}` | `WithHttpMethodsToRetry` | `HttpMethodsToRetry()` |

The retryConfiguration can be initialized as follows:

```go
retryConfiguration := verizon.CreateRetryConfiguration(
    verizon.WithMaxRetryAttempts(0),
    verizon.WithRetryOnTimeout(true),
    verizon.WithRetryInterval(1),
    verizon.WithMaximumRetryWaitTime(0),
    verizon.WithBackoffFactor(2),
    verizon.WithHttpStatusCodesToRetry([]int64{408, 413, 429, 500, 502, 503, 504, 521, 522, 524}),
    verizon.WithHttpMethodsToRetry([]string{"GET", "PUT"}),
)
```



Verizon-GO_GENERIC_LIB/doc/models/dto-device-config.md

# Dto Device Config

## Structure

`DtoDeviceConfig`

## Fields

| Name | Type | Tags | Description |
|  --- | --- | --- | --- |
| `Ble` | [`*models.SensorInsightsBle`](../../doc/models/sensor-insights-ble.md) | Optional | Property objects for Bluetooth Low-Energy (BLE) devices |

## Example (as JSON)

```json
{
  "ble": {
    "dataMode": 216,
    "manufacturerId": 180,
    "maxNumScan": 126,
    "minSigStr": 60,
    "monitorPeriod": 88,
    "exampleAdditionalProperty": {
      "key1": "val1",
      "key2": "val2"
    }
  }
}
```



Verizon-CS_NET_STANDARD_LIB/Verizon-PYTHON_GENERIC_LIB/verizon/configuration.py
# -*- coding: utf-8 -*-

"""
verizon

This file was automatically generated by APIMATIC v3.0 (
 https://www.apimatic.io ).
"""

from enum import Enum
from apimatic_core.http.configurations.http_client_configuration import HttpClientConfiguration
from apimatic_requests_client_adapter.requests_client import RequestsClient


class Environment(Enum):
    """An enum for SDK environments"""
    PRODUCTION = 0
    MOCK_SERVER_FOR_LIMITED_AVAILABILITY_SEE_QUICK_START = 1


class Server(Enum):
    """An enum for API servers"""
    EDGE_DISCOVERY = 0
    IMP_SERVER = 1
    THINGSPACE = 2
    OAUTH_SERVER = 3
    M2M = 4
    DEVICE_LOCATION = 5
    SUBSCRIPTION_SERVER = 6
    SOFTWARE_MANAGEMENT_V1 = 7
    SOFTWARE_MANAGEMENT_V2 = 8
    SOFTWARE_MANAGEMENT_V3 = 9
    PERFORMANCE = 10
    DEVICE_DIAGNOSTICS = 11
    CLOUD_CONNECTOR = 12
    HYPER_PRECISE_LOCATION = 13
    SERVICES = 14
    QUALITY_OF_SERVICE = 15


class Configuration(HttpClientConfiguration):
    """A class used for configuring the SDK by a user.
    """

    @property
    def environment(self):
        return self._environment

    @property
    def thingspace_oauth_credentials(self):
        return self._thingspace_oauth_credentials

    @property
    def vz_m2m_token_credentials(self):
        return self._vz_m2m_token_credentials

    @property
    def thingspace_oauth_1_credentials(self):
        return self._thingspace_oauth_1_credentials

    @property
    def session_token_credentials(self):
        return self._session_token_credentials

    def __init__(self, http_client_instance=None,
                 override_http_client_configuration=False, http_call_back=None,
                 timeout=30, max_retries=0, backoff_factor=2,
                 retry_statuses=None, retry_methods=None,
                 logging_configuration=None, environment=Environment.PRODUCTION,
                 thingspace_oauth_credentials=None,
                 vz_m2m_token_credentials=None,
                 thingspace_oauth_1_credentials=None,
                 session_token_credentials=None):
        if retry_methods is None:
            retry_methods = ['GET', 'PUT']

        if retry_statuses is None:
            retry_statuses = [408, 413, 429, 500, 502, 503, 504, 521, 522, 524]

        super().__init__(http_client_instance,
                         override_http_client_configuration, http_call_back,
                         timeout, max_retries, backoff_factor, retry_statuses,
                         retry_methods, logging_configuration)

        # Current API environment
        self._environment = environment

        # The object holding OAuth 2 Client Credentials Grant credentials
        self._thingspace_oauth_credentials = thingspace_oauth_credentials

        # The object holding Custom Header Signature credentials
        self._vz_m2m_token_credentials = vz_m2m_token_credentials

        # The object holding OAuth 2 Client Credentials Grant credentials
        self._thingspace_oauth_1_credentials = thingspace_oauth_1_credentials

        # The object holding Custom Header Signature credentials
        self._session_token_credentials = session_token_credentials

        # The Http Client to use for making requests.
        self.set_http_client(self.create_http_client())

    def clone_with(self, http_client_instance=None,
                   override_http_client_configuration=None, http_call_back=None,
                   timeout=None, max_retries=None, backoff_factor=None,
                   retry_statuses=None, retry_methods=None,
                   logging_configuration=None, environment=None,
                   thingspace_oauth_credentials=None,
                   vz_m2m_token_credentials=None,
                   thingspace_oauth_1_credentials=None,
                   session_token_credentials=None):
        http_client_instance = http_client_instance or self.http_client_instance
        override_http_client_configuration = override_http_client_configuration or self.override_http_client_configuration
        http_call_back = http_call_back or self.http_callback
        timeout = timeout or self.timeout
        max_retries = max_retries or self.max_retries
        backoff_factor = backoff_factor or self.backoff_factor
        retry_statuses = retry_statuses or self.retry_statuses
        retry_methods = retry_methods or self.retry_methods
        logging_configuration = logging_configuration or self.logging_configuration
        environment = environment or self.environment
        thingspace_oauth_credentials = thingspace_oauth_credentials or self.thingspace_oauth_credentials
        vz_m2m_token_credentials = vz_m2m_token_credentials or self.vz_m2m_token_credentials
        thingspace_oauth_1_credentials = thingspace_oauth_1_credentials or self.thingspace_oauth_1_credentials
        session_token_credentials = session_token_credentials or self.session_token_credentials
        return Configuration(
            http_client_instance=http_client_instance,
            override_http_client_configuration=override_http_client_configuration,
            http_call_back=http_call_back, timeout=timeout,
            max_retries=max_retries, backoff_factor=backoff_factor,
            retry_statuses=retry_statuses, retry_methods=retry_methods,
            logging_configuration=logging_configuration, environment=environment,
            thingspace_oauth_credentials=thingspace_oauth_credentials,
            vz_m2m_token_credentials=vz_m2m_token_credentials,
            thingspace_oauth_1_credentials=thingspace_oauth_1_credentials,
            session_token_credentials=session_token_credentials
        )

    def create_http_client(self):
        return RequestsClient(
            timeout=self.timeout, max_retries=self.max_retries,
            backoff_factor=self.backoff_factor, retry_statuses=self.retry_statuses,
            retry_methods=self.retry_methods,
            http_client_instance=self.http_client_instance,
            override_http_client_configuration=self.override_http_client_configuration,
            response_factory=self.http_response_factory
        )

    # All the environments the SDK can run in
    environments = {
        Environment.PRODUCTION: {
            Server.EDGE_DISCOVERY: 'https://5gedge.verizon.com/api/mec/eds',
            Server.IMP_SERVER: 'https://imp.thingspace.verizon.com',
            Server.THINGSPACE: 'https://thingspace.verizon.com/api',
            Server.OAUTH_SERVER: 'https://thingspace.verizon.com/api/ts/v1',
            Server.M2M: 'https://thingspace.verizon.com/api/m2m',
            Server.DEVICE_LOCATION: 'https://thingspace.verizon.com/api/loc/v1',
            Server.SUBSCRIPTION_SERVER: 'https://thingspace.verizon.com/api/subsc/v1',
            Server.SOFTWARE_MANAGEMENT_V1: 'https://thingspace.verizon.com/api/fota/v1',
            Server.SOFTWARE_MANAGEMENT_V2: 'https://thingspace.verizon.com/api/fota/v2',
            Server.SOFTWARE_MANAGEMENT_V3: 'https://thingspace.verizon.com/api/fota/v3',
            Server.PERFORMANCE: 'https://5gedge.verizon.com/api/mec',
            Server.DEVICE_DIAGNOSTICS: 'https://thingspace.verizon.com/api/diagnostics/v1',
            Server.CLOUD_CONNECTOR: 'https://thingspace.verizon.com/api/cc/v1',
            Server.HYPER_PRECISE_LOCATION: 'https://thingspace.verizon.com/api/hyper-precise/v1',
            Server.SERVICES: 'https://5gedge.verizon.com/api/mec/services',
            Server.QUALITY_OF_SERVICE: 'https://thingspace.verizon.com/api/m2m/v1/devices'
        },
        Environment.MOCK_SERVER_FOR_LIMITED_AVAILABILITY_SEE_QUICK_START: {
            Server.EDGE_DISCOVERY: 'https://mock.thingspace.verizon.com/api/mec/eds',
            Server.IMP_SERVER: 'https://mock.thingspace.verizon.com',
            Server.THINGSPACE: 'https://mock.thingspace.verizon.com/api',
            Server.OAUTH_SERVER: 'https://mock.thingspace.verizon.com/api/ts/v1',
            Server.M2M: 'https://mock.thingspace.verizon.com/api/m2m',
            Server.DEVICE_LOCATION: 'https://mock.thingspace.verizon.com/api/loc/v1',
            Server.SUBSCRIPTION_SERVER: 'https://mock.thingspace.verizon.com/api/subsc/v1',
            Server.SOFTWARE_MANAGEMENT_V1: 'https://mock.thingspace.verizon.com/api/fota/v1',
            Server.SOFTWARE_MANAGEMENT_V2: 'https://mock.thingspace.verizon.com/api/fota/v2',
            Server.SOFTWARE_MANAGEMENT_V3: 'https://mock.thingspace.verizon.com/api/fota/v3',
            Server.PERFORMANCE: 'https://mock.thingspace.verizon.com/api/mec',
            Server.DEVICE_DIAGNOSTICS: 'https://mock.thingspace.verizon.com/api/diagnostics/v1',
            Server.CLOUD_CONNECTOR: 'https://mock.thingspace.verizon.com/api/cc/v1',
            Server.HYPER_PRECISE_LOCATION: 'https://mock.thingspace.verizon.com/api/hyper-precise/v1',
            Server.SERVICES: 'https://mock.thingspace.verizon.com/api/mec/services',
            Server.QUALITY_OF_SERVICE: 'https://mock.thingspace.verizon.com/api/m2m/v1/devices'
        }
    }

    def get_base_uri(self, server=Server.EDGE_DISCOVERY):
        """Generates the appropriate base URI for the environment and the
        server.

        Args:
            server (Configuration.Server): The server enum for which the base
            URI is required.

        Returns:
            String: The base URI.

        """
        return self.environments[self.environment][server]
